### ISSUE TITLE
Kafka doesn’t work with SSL

### ISSUE THREAD

**Issue by luminorena** [mostafa/xk6-kafka#212]
Hello, @mostafa!

I use xk6-kafka library to perform load testing using Apache Kafka and SSL. I connect to production Kafka easily (it works with JKS certificates through Kafka Tool).  However, I cannot connect to production Kafka via k6 getting strange error “Go error: Failed to get controller: %!w(*errors.errorString=&{EOF})” pointing on this line:

 ```javascript
  const connection = new Connection({
  address: brokers[0],
  tlsConfig,
  });

```

I also have this error: “level=error msg=”Cannot process TLS config” error=”File not found”. However, I have tried to load certificates and keys from a JKS keystore using test_tls_with_jks.js script taken from your examples. And it works correctly without any errors on production. Therefore, I suppose there is no problem with certificates. 
Here you can see the code I use (I have the same key for keystore/truststore):

```javascript
import { check } from "k6";
import {
  Writer,
  Reader,
  Connection,
  SchemaRegistry,
  CODEC_SNAPPY,
  SCHEMA_TYPE_JSON,
} from "k6/x/kafka";

import { LoadJKS, TLS_1_2 } from "k6/x/kafka";
const jks = LoadJKS({
  path: "kafka.keystore.jks",
  password: "password",
  clientCertAlias: "localhost",
  clientKeyAlias: "localhost",
  clientKeyPassword: "password",
  serverCaAlias: "localhost",
  });

const tlsConfig = {
  enableTls: true,
  insecureSkipTlsVerify: false,
  minVersion: TLS_1_2,
  clientCertPem: jks["clientCertsPem"][0],
  clientKeyPem: jks["clientKeyPem"],
  serverCaPem: jks["serverCaPem"],
};

const brokers = ["localhost:9092"];
const topic = "test_1";

const writer = new Writer({
  brokers: brokers,
  tlsConfig,
  topic: topic,
  autoCreateTopic: true,
  compression: CODEC_SNAPPY,
});

const reader = new Reader({
  brokers: brokers,
  tlsConfig,
  topic: topic,
});

const connection = new Connection({
  address: brokers[0],
  tlsConfig,
});

const schemaRegistry = new SchemaRegistry();

if (__VU == 0) {
  connection.createTopic({
    topic: topic,
    configEntries: [
      {
        configName: "compression.type",
        configValue: CODEC_SNAPPY,
      },
    ],
  });
}

export const options = {
  thresholds: {
    kafka_writer_error_count: ["count == 0"],
    kafka_reader_error_count: ["count == 0"],
  },
};

function getRandomInt(min, max) {
  return min + Math.floor(Math.random() * (max - min + 1));
}

export default function () {
  let messages = [
    {
      key: schemaRegistry.serialize({
        data: {
          correlationId: "test-id-abc-" + getRandomInt(10, 100),
        },
        schemaType: SCHEMA_TYPE_JSON,
      }),
      value: schemaRegistry.serialize({
        data: {
          menu: {
            version: "test_1",
            author: "test_2",
            description: "test_3",
          }

        },
        schemaType: SCHEMA_TYPE_JSON,
      })

    }
  ];

  writer.produce({ messages: messages });

  messages = reader.consume({});

}

export function teardown(data) {
  console.log(Object.keys(jks));
  console.log(jks);
  console.log(tlsConfig);

  writer.close();
  reader.close();
  connection.close();
}

```

_What I have also tried:_
1. I have used two different Kafka instances to eliminate problem with certificates. 
2. I have installed docker on my local machine and tried to connect to Kafka (without SSL). As a result, I have sent a successful message to my local Kafka.
3. I have tried different versions of TLS but nothing worked.
4. I have tried to transform JKS certificates to PEM but my production Kafka does not work with them. It needs only pure JKS certs.

I have looked through some of your issues and I have seen one possible solution to use SASL SSL connection to Kafka. But it is not possible in my case. I have only JKS certificates to connect to production Kafka.

Is there a work-around to my problem ?

**Comment by mostafa** [comment]
Hey @luminorena,

now that I look once more at the code, I found an issue. Although the [JKS](https://github.com/mostafa/xk6-kafka/blob/main/scripts/test_tls_with_jks.js) script feels correct, but the `tlsConfig` expects file paths, instead of entire keys and certificates as string. I will fix this as soon as I can.

**Comment by mostafa** [comment]
Hey @luminorena,

I updated the `LoadJKS` implementation to save the certificates and keys in the present working directory and return the filenames instead of stringified certificates and keys. Will you test it to see if it works for you?

**Comment by luminorena** [comment]
Hi @mostafa,

That's awesome! I'll definitely check it these days because our project really needs this implementation. 
When it's done I'll make a comment here.

**Comment by mostafa** [comment]
@luminorena 

Just make sure that you build xk6-kafka for [development](https://github.com/mostafa/xk6-kafka#build-for-development), since I haven't released a version yet. I'll release a version after your tests are successful.

**Comment by luminorena** [comment]
Hi @mostafa, 

I've tested a new version. Before doing that, I've built xk6-kafka for development. No code has been previously changed. 
I used jks certificate we have to connect to our Kafka instance. Your script has successfully generated some pem certs, but I could not connect to our Kafka because of the error:

Cannot process TLS config error="Error creating x509 key pair from client-cert-0.pem and client-key.pem, Original error: %!w(*errors.errorString=&{tls: failed to find any PEM data in certificate input})"

If I run pure test_tls_with_jks.js script from your examples using our certificate, there won't be any errors. However, if I try to connect to Kafka instance I get this error (the code I use is in the first message of this issue). 

I've also tried different Kafka instances that work with TLS on production, but the error is the same. 

Am I doing eveything right?

**Comment by mostafa** [comment]
@luminorena 

I look under the hood, and found out that the generated files are in binary format (DER), which needs to be encoded to PEM format. So, I made a new commit (https://github.com/mostafa/xk6-kafka/commit/915fb6b9cabd543566d7d7c364193975c0bde184) to do just that. Could you please test it again?

**Comment by luminorena** [comment]
@mostafa

We don't use PKCS#12 format. 
After your last commit everything works pefrectly!  I can send message to topic with k6 and visualize it with Kafka Tool. 👍 
Now you can release a new version.

Thanks!

**Comment by mostafa** [comment]
@luminorena 

Thanks a lot for testing this. 🙏 I'll go for a release! 🚀

**Comment by mostafa** [comment]
I just released [v0.18.0](https://github.com/mostafa/xk6-kafka/releases/tag/v0.18.0).
